dashboard/test/ui/features/step_definitions/steps.rb

```diff
@@ -319,6 +319,10 @@ def jquery_is_element_open(selector)
   sleep seconds.to_f
 end
 
+When /^I wait for the video thumbnails to load$/ do
+  wait_until {@browser.execute_script("return $('.video-thumbnail-loaded').length == $('.video-thumbnail-wrapper').length")}
```

Comment:
1. You should probably also wait until the page itself loads because, before it does, there will be 0 of each and this will pass.
2. You can probably just write the query to look for an empty set of loading images. Something like:

```
Array.from(document.querySelectorAll('img.thumbnail-image')).filter((img) => !img.complete).length == 0
```

Reply:
I'm not sure exactly which steps function you would mean by #1. I initially looked at 'and I wait for the page to fully load', but that step looks for the run button, and all these are various marketing pages. Additionally, because they are all different pages (the feature file uses a template), they don't really have one thing in common I could wait for. Thankfully, they all have a language selector that has to load and seems to pop up after the page starts to load which buys each test enough time to have the empty video containers (at the very least).

Reply:
Can you say more about img.complete? I went with this approach because it ensures we have a loaded image for every video container, but I could see an advantage to this filter approach if I could utilize some sort of inherent image attribute instead of having to insert it myself in our display_video_thumbnail.haml file.

Reply:
The `img` element has a `complete` property that is `true` when the image is loaded. I think `script` tags have this too: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/complete

Maybe check for at least one video thumbnail! The step assumes they exist, that's a cheap way to do it. The images won't load before the page loads.

```
document.querySelector('img.thumbnail-image') && Array.from(document.querySelectorAll('img.thumbnail-image')).filter((img) => !img.complete).length == 0
```

Reply:
I think testing for the existence of the `footer` is also a reasonable way of knowing the entire structure of the page is there.

```
document.querySelector('footer') && Array.from(document.quer...
```

Reply:
I'm not sure all the pages have videos, so was hoping to write this in a way that it would work regardless (0 == 0). But I love the idea of checking for the footer, and will look into img.complete! Thanks!

Reply:
> 1. You should probably also wait until the page itself loads because, before it does, there will be 0 of each and this will pass.

I think we are good here without checking the footer, because the videos we are trying to deflake are not rendered with react. the `I am on` step will wait for the document ready state, here: https://github.com/code-dot-org/code-dot-org/blob/bef817e3d1d3d47e21ad0ee72d8f7859ea64996b/dashboard/test/ui/features/step_definitions/steps.rb#L88

> 2. You can probably just write the query to look for an empty set of loading images. Something like:
> 
> ```
> Array.from(document.querySelectorAll('img.thumbnail-image')).filter((img) => !img.complete).length == 0
> ```

🤯 wow, yes, the img.complete option looks great!

Reply:
Yep. That there is a step that already waits for the page to load is good enough! Coming back to it fresh, I think the `img.complete` option is going to be the least intrusive and simplest way to check for this without adding anything test-related to the pages themselves.